core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/TagQueryParser.java

```diff
@@ -0,0 +1,243 @@
+        }
+
+        return groupMetrics;
+    }
+
+    public Observable<Map<String, Set<String>>> getTagValues(String tenantId, MetricType<?> metricType,
+                                                             Map<String, String> tagsQueries) {
+
+        // Row: 0 = type, 1 = metricName, 2 = tagValue, e.getKey = tagName, e.getValue = regExp
+        return Observable.from(tagsQueries.entrySet())
+                .flatMap(e -> dataAccess.findMetricsByTagName(tenantId, e.getKey())
+                        .filter(typeFilter(metricType))
+                        .filter(tagValueFilter(e.getValue(), 3))
+                        .map(row -> {
+                            Map<String, Map<String, String>> idMap = new HashMap<>();
+                            Map<String, String> valueMap = new HashMap<>();
+                            valueMap.put(e.getKey(), row.getString(3));
+
+                            idMap.put(row.getString(2), valueMap);
+                            return idMap;
+                        })
+                        .switchIfEmpty(Observable.just(new HashMap<>()))
+                        .reduce((map1, map2) -> {
+                            map1.putAll(map2);
+                            return map1;
+                        }))
+                .reduce((m1, m2) -> {
+                    // Now try to emulate set operation of cut
+                    Iterator<Map.Entry<String, Map<String, String>>> iterator = m1.entrySet().iterator();
+
+                    while (iterator.hasNext()) {
+                        Map.Entry<String, Map<String, String>> next = iterator.next();
+                        if (!m2.containsKey(next.getKey())) {
+                            iterator.remove();
+                        } else {
+                            // Combine the entries
+                            Map<String, String> map2 = m2.get(next.getKey());
+                            map2.forEach((k, v) -> next.getValue().put(k, v));
+                        }
+                    }
+
+                    return m1;
+                })
+                .map(m -> {
+                    Map<String, Set<String>> tagValueMap = new HashMap<>();
+
+                    m.forEach((k, v) ->
+                            v.forEach((subKey, subValue) -> {
+                                if (tagValueMap.containsKey(subKey)) {
+                                    tagValueMap.get(subKey).add(subValue);
+                                } else {
+                                    Set<String> values = new HashSet<>();
+                                    values.add(subValue);
+                                    tagValueMap.put(subKey, values);
+                                }
+                            }));
+
+                    return tagValueMap;
+                });
+    }
+
+    private Func1<Metric<?>, Boolean> tagNotExistsFilter(String unwantedTagName) {
+        return tMetric -> {
```

Comment:
Yeah, there used to be more functionality in it, but after removing those it would have made sense to refactor it.